pkg/cable/wireguard/driver.go

```diff
@@ -0,0 +1,411 @@
+func (w *wireguard) DisconnectFromEndpoint(remoteEndpoint types.SubmarinerEndpoint) error {
+	klog.V(log.DEBUG).Infof("Removing endpoint %v+", remoteEndpoint)
+
+	if w.localEndpoint.Spec.ClusterID == remoteEndpoint.Spec.ClusterID {
+		klog.V(log.DEBUG).Infof("Will not disconnect self")
+		return nil
+	}
+
+	// parse remote public key
+	remoteKey, err := publicKey(remoteEndpoint)
+	if err != nil {
+		return fmt.Errorf("failed to parse peer public key: %v", err)
+	}
+
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	// key mismatch indicates removal of old peer key, where clusterID is already associated with new key or no key
+	keyMismatch := false
+
+	oldKey, found := w.peers[remoteEndpoint.Spec.ClusterID]
+	if !found {
+		keyMismatch = true
+		klog.Warningf("Key mismatch, cluster %s has no key but asked to remove %s", remoteEndpoint.Spec.ClusterID, remoteKey)
+	} else if oldKey.String() != remoteKey.String() {
+		keyMismatch = true
+		klog.Warningf("Key mismatch, cluster %s key is %s but asked to remove %s", remoteEndpoint.Spec.ClusterID, oldKey, remoteKey)
+	}
+
+	// wg remove
+	_ = removePeer(w.client, remoteKey)
+
+	if keyMismatch {
+		// ClusterID probably already associated with new key. Do not remove peers entry nor routes
+		klog.Warningf("Key mismatch for peer cluster %s, keeping existing routes and public key", remoteEndpoint.Spec.ClusterID)
+		return nil
+	}
+	// no delete on mismatch
+	delete(w.peers, remoteEndpoint.Spec.ClusterID)
+
+	// del routes
+	allowedIPs := parseSubnets(remoteEndpoint.Spec.Subnets)
+	idx := w.link.Attrs().Index
+	for _, peerNet := range allowedIPs {
+		route := netlink.Route{
+			LinkIndex: idx,
+			Dst:       &peerNet,
+		}
+		if err = netlink.RouteDel(&route); err != nil {
+			return fmt.Errorf("failed to delete route %s: %v", route, err)
+		}
+	}
+
+	klog.V(log.DEBUG).Infof("Done removing endpoint for cluster %s", remoteEndpoint.Spec.ClusterID)
+	return nil
+}
+
+func (w *wireguard) GetActiveConnections(clusterID string) ([]string, error) {
+	// force caller to skip duplicate handling
+	return make([]string, 0), nil
+}
+
+// Create new wg link and assign addr from local subnets
+func setWGLink(w *wireguard, localSubnets []string) error {
+
+	// delete existing wg device if needed
+	if link, err := netlink.LinkByName(DefaultDeviceName); err == nil {
+		// delete existing device
+		if err := netlink.LinkDel(link); err != nil {
+			return fmt.Errorf("failed to delete existing WireGuard device: %v", err)
+		}
+	}
+
+	// create the wg device (ip link add dev $DefaultDeviceName type wireguard)
+	la := netlink.NewLinkAttrs()
+	la.Name = DefaultDeviceName
+	link := &netlink.GenericLink{
+		LinkAttrs: la,
+		LinkType:  "wireguard",
+	}
+	if err := netlink.LinkAdd(link); err == nil {
+		w.link = link
+	} else {
+		return fmt.Errorf("failed to add WireGuard device: %v", err)
+	}
+
+	// parse localSubnets and get internal address
+	w.localSubnets = parseSubnets(localSubnets)
+	ip, err := discoverInternalIP(w.localSubnets)
```

Comment:
What if `discoverInternalIP` returned an error indicating it wasn't able to determine if an internal address exists? Should we still proceed with using the endpoint IP?

Reply:
that is what is attempted at L311. 
Using endpoint IP will probably cause problems for pods on host network.

Reply:
L311 is a log statement - did you mean L314? 😄 The call site here can't distinguish between an internal address not found and an error trying to determine if an internal address exists from `discoverInternalIP`.  For the former,  we certainly want to proceed to L309 to use the endpoint IP.  I'm unclear about the latter but if we still want to proceed with attempting to use the endpoint IP in this case as well then I think we should at least log the error. I propose to return nil from `discoverInternalIP` if no address is found and log the returned error here, ie:

```
ip, err := discoverInternalIP(w.localSubnets)
if err != nil {
    klog.Errorf("Error attempting to discover an internal address: %v", err)
}

if ip == nil {
    klog.V(log.DEBUG).Infof("Using endpoint IP as internal address; %v", err)
    ip = endpointIP(&w.localEndpoint)
}

```

Reply:
Actually L309 (removed 2 comments above...).
I see your point; however, `discoverInternalIP` returns `""` in both cases.
I can change `discoverInternalIP` to only log the 'not found' case, but I still see both cases as error, and still think the best thing we can do is try the ep IP.

Reply:
> Actually L309 (removed 2 comments above...).
> I see your point; however, `discoverInternalIP` returns `""` in both cases.

yeah, sorry, return `""` instead of nil.

> I can change `discoverInternalIP` to only log the 'not found' case, but I still see both cases as error, and still think the best thing we can do is try the ep IP.

yeah try the endpoint IP in both cases - I'm just suggesting to log the error if something failed preventing the determination so we don't lose it in case it might be helpful for troubleshooting. So the last line in `discoverInternalIP`  where it's not found  would return  `"", nil`.

Reply:
done